### PR TITLE
Do not flag fields with certain special attributes as unused (…

### LINKED ISSUE #1217: Remove unused field (CA1823) reported for MEF exported fields
```
[Export]
private readonly int _field;
```

This reports CA1823, which is a false positive.

### LINKED ISSUE #1219: Remove unused field (CA1823) reported for fields marked with FieldOffsetAttribute
```
[FieldOffset(8)]
private readonly long _longValue;
```

We should not fire for fields with this attribute.

### PR DESCRIPTION
**Pull request by mavasani**
…CA1823)

Fixes #1219
Fixes #1217

### CODE REVIEW

File: src/Microsoft.CodeQuality.Analyzers/Core/Maintainability/AvoidUnusedPrivateFields.cs [hunk 1]
```diff
@@ -48,12 +48,27 @@ public override void Initialize(AnalysisContext analysisContext)
...
+                    
                     compilationContext.RegisterSymbolAction(
                         (symbolContext) =>
                         {
                             IFieldSymbol field = (IFieldSymbol)symbolContext.Symbol;
                             if (field.DeclaredAccessibility == Accessibility.Private && !referencedPrivateFields.Contains(field))
                             {
+                                if (!specialAttributes.IsEmpty)
+                                {
+                                    var fieldAttributes = field.GetAttributes();
```
Comment (333fred) [inline src/Microsoft.CodeQuality.Analyzers/Core/Maintainability/AvoidUnusedPrivateFields.cs]: Unused?

Comment (mavasani) [reply src/Microsoft.CodeQuality.Analyzers/Core/Maintainability/AvoidUnusedPrivateFields.cs]: Yes, how ironic :-) We should really enable analyzers again on our own repo.

File: src/Microsoft.CodeQuality.Analyzers/UnitTests/Maintainability/AvoidUnusedPrivateFieldsTests.cs [hunk 2]
```diff
@@ -109,6 +212,76 @@ End Class
...
+    Private fieldWithMefV1ExportAttribute As Integer
+
+    <System.ComponentModel.Composition.ExportAttribute(0)> _
+    Private fieldWithMefV2ExportAttribute As Integer
+End Class
+", TestValidationMode.AllowCompileErrors);
+        }
+
+        [Fact, WorkItem(1217, "https://github.com/dotnet/roslyn-analyzers/issues/1217")]
+        public void CA1823_VisualBasic_MEFAttributesUndefined_Diagnostic()
```
Comment (333fred) [inline src/Microsoft.CodeQuality.Analyzers/UnitTests/Maintainability/AvoidUnusedPrivateFieldsTests.cs]: No versions of this test for FieldOffsetAttribute?

**Comment by 333fred** [comment]
Do we also want to address @sharwell's comment on `LayoutKind.Sequential`, [https://github.com/dotnet/roslyn-analyzers/issues/1219#issuecomment-305309494](https://github.com/dotnet/roslyn-analyzers/issues/1219#issuecomment-305309494).

File: src/Microsoft.CodeQuality.Analyzers/UnitTests/Maintainability/AvoidUnusedPrivateFieldsTests.cs (hunk 2 above)
Comment (mavasani) [reply src/Microsoft.CodeQuality.Analyzers/UnitTests/Maintainability/AvoidUnusedPrivateFieldsTests.cs]: FieldOffsetAttribute is defined in mscorlib.

**Comment by mavasani** [comment]
> Do we also want to address @sharwell's comment on LayoutKind.Sequential

Yes, let me do so while I am touching this code.

**Comment by mavasani** [comment]
Ping.. any other feedback?

**Pull request merged**